src/ast/diagnostic.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <cassert>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "location.hh"
+
+namespace bpftrace {
+namespace ast {
+
+// Diagnostic reflects a single error at a single source location. This is a
+// simple wrapper around a string for that message, and the location class.
+class Diagnostic {
+public:
+  enum class Severity {
+    Warning,
+    Error,
+  };
+  Diagnostic(location loc) : loc_(loc) {};
+  const std::string msg() const
+  {
+    return msg_.str();
+  }
+  const std::string hint() const
+  {
+    return hint_.str();
+  }
+  const location& loc() const
+  {
+    return loc_;
+  }
+
+  // Each diagnostic can potentially have a hint attached, which is how to
+  // effectively resolve this issue.
+  std::stringstream& addHint()
```

Comment:
Judging by the name, shouldn't this write to `hint_`?

Reply:
Is it silly for this function to just accept a string and add it to the stream itself?

Reply:
The existing code for adding a hint uses a stream, and a stream makes the most sense. Otherwise I would be reworking that code to use a `std::stringstream` and then making a call to add hint?

See the subsequent changes:
<img width="1337" alt="Screenshot 2025-02-25 at 2 27 45 PM" src="https://github.com/user-attachments/assets/61f5f1c6-8df7-4437-9e6b-7429c134fe4c" />

Reply:
Yeah, stream makes sense. I can see now that this is consistent with `addError()` etc. below so let's keep `addHint()` (even though it still reads a bit weird to me).